python/lsst/daf/butler/registries/sqlRegistry.py

```diff
@@ -523,81 +523,39 @@ def associate(self, collection, refs):
                 ref.datasetType._dimensions = self.dimensions.extract(ref.datasetType.dimensions)
 
-            dataId = ref.dataId
-            if all(row[col] == dataId[col] for col in ref.datasetType.dimensions.links()):
-                raise ValueError("A dataset of type {} with id: {} already exists in collection {}".format(
-                    ref.datasetType, dataId, collection))
-            return False
-
-        if len(refs) == 1:
-            # small optimization for a single ref
-            ref = refs[0]
-            dataset_id = self._findDatasetId(collection, ref.datasetType, ref.dataId)
-            if dataset_id == ref.id:
-                # already there
-                return
-            elif dataset_id is not None:
-                raise ValueError("A dataset of type {} with id: {} already exists in collection {}".format(
-                    ref.datasetType, ref.dataId, collection))
-        else:
-            # full scan of a collection to compare DatasetRef dimensions
-            datasetTable = self._schema.tables["Dataset"]
-            datasetCollectionTable = self._schema.tables["DatasetCollection"]
-            query = datasetTable.select()
-            query = query.where(and_(datasetTable.c.dataset_id == datasetCollectionTable.c.dataset_id,
-                                     datasetCollectionTable.c.collection == collection))
-            result = self._connection.execute(query)
-            for row in result:
-                # skip DatasetRefs that are already there
-                refs = [ref for ref in refs if not _matchRef(row, ref)]
-
-        # if any ref is not there yet add it
-        if refs:
-            datasetCollectionTable = self._schema.tables["DatasetCollection"]
-            self._connection.execute(datasetCollectionTable.insert(),
-                                     [{"dataset_id": ref.id, "dataset_ref_hash": ref.hash,
-                                       "collection": collection} for ref in refs])
+            row = self._connection.execute(query, hash=ref.hash).fetchone()
+            if row is None:
+                # No Dataset with this DatasetType and Data ID in collection;
+                # insert it now.
+                self._connection.execute(datasetCollectionTable.insert(),
+                                         [{"dataset_id": ref.id, "dataset_ref_hash": ref.hash,
+                                           "collection": collection} for ref in refs])
+            elif row.dataset_id != ref.id:
+                # A different Dataset with this DatasetType and Data ID already
+                # exist in this collection.
+                raise ValueError(
+                    "A dataset of type {} with id: {} already exists in collection {}".format(
+                        ref.datasetType, ref.dataId, collection
+                    )
+                )
+            # If the same Dataset is already in this collection, do nothing.
```

Comment:
Interesting - I think this indicates a fundamental misunderstanding by me on what transactions guarantee in general, which means I need to spend some time thinking about what those guarantees actually are.

I don't think "analyzing the `DuplicateKey` error" approach can really be done portably, either, because  we don't have any guarantee about what kind of information the error message might contain (and we need to know which constraint failed), so ultimately I think this (and several other `Registry` methods) need to be left pure-abstract in `SqlRegistry` so they must be implemented by DB-specific derived classes.

As I said on Jira, I think it's best not to do that now because (for now) a solution that's subject to race conditions is better than no solution at all.  I'll be sure to time the future changes carefully with NCSA to make sure we don't have a regression in Oracle support at an awkward time.

Reply:
By analyzing I meant that you would need to select from database at that point, after failure. Relying on exception content is not possible of course.